Ke team suggestions (#26)
cqframework/CQL-Formatting-and-Usage-Wiki: Source/CheatSheet/cheatsheet.md

6) [Value Sets & Code Systems](https://cql.hl7.org/02-authorsguide.html#terminology) - Value set and code system declarations allow terminology to be referenced anywhere within the library.

```cql

> include valueset "Acute Pharyngitis": 'http://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113883.3.464.1003.102.12.1011'
> include codesystem "SNOMED": 'http://snomed.info/sct'

Reviewer: No need for `include` keyword here
```suggestion
valueset "Acute Pharyngitis": 'http://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113883.3.464.1003.102.12.1011'
codesystem "SNOMED": 'http://snomed.info/sct'
```

2) [terminology filter part](https://cql.hl7.org/02-authorsguide.html#filtering-with-terminology) : the retrieve expression allows the results to be filtered using terminology, including value sets, code systems, or by specifying a single code.

- The example below filters the results to only Conditions whose code is in the "Acute Pharyngitis" value set.

```cql
[Condition: "Acute Pharyngitis"]
```

- The example below returns Conditions with class in the "Inpatient Encounters" value set.

```cql
valueset "Inpatient Encounters": 'http://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113883.3.666.5.307'
[Condition: class in "Inpatient Encounters"]
```

- The example below retrieves Conditions with codes equivalent to the "Diabetes Code" direct-reference code.

```cql
code "Diabetes Code": '123' from "LOINC"
[Condition: code ~ "Diabetes Code"]
```

### [**Function Syntax**](https://cql.hl7.org/19-l-cqlsyntaxdiagrams.html#function)

> A function in CQL is a named expression that is allowed to take any number of arguments. A function can be invoked directly by name. A colon must end the quoted identifier.

Reviewer: The function name needs to be an identifier (either simple or quoted), so I would suggest removing the bit about the colon entirely, since that's true of any define expression and isn't unique to functions.

```suggestion
A function in CQL is a named expression that is allowed to take any number of arguments. A function can be invoked directly by name.
```

**Strings use single quotation marks.**

[Single Quotes](https://cql.hl7.org/19-l-cqlsyntaxdiagrams.html#string) 

Single quotes are limited to string literals
1) text
```cql
define "string literal": 'hello'
```

2) code system and value set object identifiers (OID)
```cql
valueset "example":  'urn:oid:2.16.840.1.113883.3.560.100.2'
```

3) version identifiers
```cql
valueset "example":  'urn:oid:2.16.840.1.113883.3.560.100.2' version '123'
```

4) code declarations
```cql
code "code": '123' from "codesystem identifier"
```

5) units
```cql
define "quantity": 12.0 'L'
```

**Identifiers may use double quotation marks.**

[Quoted Identifiers](https://cql.hl7.org/19-l-cqlsyntaxdiagrams.html#QUOTEDIDENTIFIER)
This applies to definitions, functions, valuesets, codes, etc... 

```cql
"SNOMED CT" // A code system declaration
"Inpatient Encounters" // A Defined Code
```

### [**Identifiers**](https://cql.hl7.org/03-developersguide.html#identifiers)
Identifiers are used to name various elements within the language. There are three types of identifiers in CQL, simple, delimited, and quoted.

```cql
Foo1 // simple
`Encounter, Performed` // delimited
"Inpatient Encounters" // quoted
```

> 

Reviewer: Suggest moving this section before the "Strings vs Identifiers" section

## Bracket Syntax

> 1. [Intervals](https://cql.hl7.org/19-l-cqlsyntaxdiagrams.html#intervalSelector) use [] and ()

Reviewer: Suggest we add narrative to describe the difference between square bracket and parentheses:

```suggestion
1. [Intervals](https://cql.hl7.org/19-l-cqlsyntaxdiagrams.html#intervalSelector) use [] and (). Following standard mathematics notation, inclusive (closed) boundaries are indicated with square brackets, and exclusive (open) boundaries are indicated with parentheses.
```

## Queries

### [**Full Query Syntax**](https://cql.hl7.org/02-authorsguide.html#full-query)

The clauses, described in the clauses section later, must appear in the correct order in order to specify a valid CQL query. The general order of clauses is:

```bnf
<primary-source> <alias>
  <with-or-without-clauses>
  <where-clause>
  <return-clause>
  <sort-clause>
```

> [link to query-clauses](#queryclauses)

Reviewer: ```suggestion
[Jump to query clauses section](#queryclauses)
```

```cql
define TestPrimitives:
  Patient P
    where P.gender.value = 'male' 
      and P.gender.value != 'female'
```

```cql
define "Absence of Cervix":
  [Procedure: "Hysterectomy with No Residual Cervix"] NoCervixProcedure
    where FC.ToInterval(NoCervixProcedure.performed) ends on or before end of "Measurement Period" 
      and NoCervixProcedure.status = 'completed'
```

### [**Type Operators**](https://cql.hl7.org/09-b-cqlreference.html#type-operators-1)

The `as` operator allows the result of an expression to be cast as a given target type

```cql
define "Former smoker observation":
  "Most recent smoking status observation" O
    where (O.value as CodeableConcept) ~ "Former Smoker"
```

The `is` operator allows the type of a result to be tested

```cql

> define "Patient is Active":
>   Patient P
>     where P.active.value is true

Reviewer: This should be testing type, not value:
```suggestion
define "Patient BirthDate is a Date":
  Patient P
    return P.birthDate is FHIR.date
```
